### ISSUE TITLE
Consul TLS configuration With terraform Genrated Certs Error

### ISSUE THREAD

**Issue by Epiclemonaid** [hashicorp/consul-helm#441]
Generating Certificates with Terraform and then adding the cakey and cacrt into helm chart leaves this error

`==> WARNING: Server Certificates grants authority to become a
    server and access all state in the cluster including root keys
    and all ACL tokens. Do not distribute them to production hosts
    that are not server nodes. Store them as securely as CA keys.
==> Using /consul/tls/ca/cert/tls.crt and /consul/tls/ca/key/tls.key
unknown PEM block type for signing key: RSA PRIVATE KEY
`
![Screenshot from 2020-04-30 10-46-25](https://user-images.githubusercontent.com/11666664/80736877-f9adc600-8acf-11ea-84ef-ff27157eccaa.png)
![Screenshot from 2020-04-30 10-48-19](https://user-images.githubusercontent.com/11666664/80736998-25c94700-8ad0-11ea-85b0-1e61068e8a84.png)

I have checked and the certificate is an `rsa private key` and is valid

**here is the terraform:**

```
resource "tls_private_key" "consul-ca" {
  algorithm = "RSA"
  rsa_bits  = "2048"
}

resource "tls_self_signed_cert" "consul-ca" {
  key_algorithm   = tls_private_key.consul-ca.algorithm
  private_key_pem = tls_private_key.consul-ca.private_key_pem

  subject {
    common_name  = "consul.example.local"
    organization = "HashiCorp Consul"
  }

  validity_period_hours = 8760
  is_ca_certificate     = true

  allowed_uses = [
    "cert_signing",
    "digital_signature",
    "key_encipherment",
  ]
}

resource "tls_private_key" "consul" {
  algorithm = "RSA"
  rsa_bits  = "2048"
}

resource "tls_cert_request" "consul" {
  key_algorithm   = tls_private_key.consul.algorithm
  private_key_pem = tls_private_key.consul.private_key_pem

  dns_names = [
    "vault",
    "consul",
    "vault.example.local",
    "consul.example.local",
    "vault.default.svc.cluster.local",
    "consul.default.svc.cluster.local",
    "localhost",
    "127.0.0.1",
  ]

  subject {
    common_name  = "consul.example.local"
    organization = "HashiCorp Consul"
  }
}

resource "tls_locally_signed_cert" "consul" {
  cert_request_pem = tls_cert_request.consul.cert_request_pem

  ca_key_algorithm   = tls_private_key.consul-ca.algorithm
  ca_private_key_pem = tls_private_key.consul-ca.private_key_pem
  ca_cert_pem        = tls_self_signed_cert.consul-ca.cert_pem

  validity_period_hours = 8760

  allowed_uses = [
    "cert_signing",
    "client_auth",
    "digital_signature",
    "key_encipherment",
    "server_auth",
  ]
}

data "google_client_config" "current" {}


provider "kubernetes" {
}


resource "kubernetes_secret" "consul-ca-cert" {
  metadata {
    name = "consul-ca-cert"
    namespace = "default"
  }

  data           = {
    "tls.crt" = tls_self_signed_cert.consul-ca.cert_pem
  }
}

resource "kubernetes_secret" "consul-ca-key" {
  metadata {
    name = "consul-ca-key"
    namespace = "default"
  }

  data           = {
    "tls.key" = tls_private_key.consul-ca.private_key_pem
  }
}
```

this runs as expected and they are valid certs according to ssl decoder
**here is my helm values** (that were alterted)

```
# global holds values that affect multiple components of the chart.
global:
  # enabled is the master enabled/disabled setting.
  # If true, servers, clients, Consul DNS and the Consul UI will be enabled.
  # Each component can override this default via its component-specific
  # "enabled" config.
  # If false, no components will be installed by default and per-component
  # opt-in is required, such as by setting `server.enabled` to true.
  enabled: true

  # name sets the prefix used for all resources in the helm chart.
  # If not set, the prefix will be "<helm release name>-consul".
  name: null

  # domain is the domain Consul will answer DNS queries for
  # (see https://www.consul.io/docs/agent/options.html#_domain) and the domain
  # services synced from Consul into Kubernetes will have,
  # e.g. `service-name.service.consul`.
  domain: consul

  # image is the name (and tag) of the Consul Docker image for clients and
  # servers. This can be overridden per component.
  # This should be pinned to a specific version tag, otherwise you may
  # inadvertently upgrade your Consul version.
  #
  # Examples:
  #   # Consul 1.5.0
  #   image: "consul:1.5.0"
  #   # Consul Enterprise 1.5.0
  #   image: "hashicorp/consul-enterprise:1.5.0-ent"
  image: "consul:1.7.2"

  # array of objects containing image pull secret names that will be applied to
  # each service account.
  # This can be used to reference image pull secrets if using
  # a custom consul or consul-k8s Docker image.
  # See https://kubernetes.io/docs/concepts/containers/images/#using-a-private-registry.
  #
  # Example:
  #   imagePullSecrets:
  #   - name: pull-secret-name
  #   - name: pull-secret-name-2
  imagePullSecrets: []

  # imageK8S is the name (and tag) of the consul-k8s Docker image that
  # is used for functionality such as catalog sync. This can be overridden
  # per component.
  # Note: support for the catalog sync's liveness and readiness probes was added
  # to consul-k8s 0.6.0. If using an older consul-k8s version, you may need to
  # remove these checks to make the sync work.
  # If using acls.manageSystemACLs then must be >= 0.10.1.
  # If using connect inject then must be >= 0.10.1.
  # If using Consul Enterprise namespaces, must be >= 0.12.
  imageK8S: "hashicorp/consul-k8s:0.13.0"

  # datacenter is the name of the datacenter that the agents should register
  # as. This can't be changed once the Consul cluster is up and running
  # since Consul doesn't support an automatic way to change this value
  # currently: https://github.com/hashicorp/consul/issues/1858.
  datacenter: dc1

  # enablePodSecurityPolicies controls whether pod
  # security policies are created for the Consul components created by this
  # chart. See https://kubernetes.io/docs/concepts/policy/pod-security-policy/.
  enablePodSecurityPolicies: false

  # gossipEncryption configures which Kubernetes secret to retrieve Consul's
  # gossip encryption key from (see https://www.consul.io/docs/agent/options.html#_encrypt).
  # If secretName or secretKey are not set, gossip encryption will not be enabled.
  # The secret must be in the same namespace that Consul is installed into.
  #
  # The secret can be created by running:
  #    kubectl create secret generic consul-gossip-encryption-key \
  #      --from-literal=key=$(consul keygen).
  #
  # In this case, secretName would be "consul-gossip-encryption-key" and
  # secretKey would be "key".
  gossipEncryption:
    # secretName is the name of the Kubernetes secret that holds the gossip
    # encryption key. The secret must be in the same namespace that Consul is installed into.
    #secretName: "consul-gossip-encryption-key"
    ## secretKey is the key within the Kubernetes secret that holds the gossip
    ## encryption key.
    #secretKey: "key"
    secretName: "consul"
    # secretKey is the key within the Kubernetes secret that holds the gossip
    # encryption key.
    secretKey: "gossip-encryption-key"

  # Enables TLS encryption across the cluster to verify authenticity of the
  # servers and clients that connect. Note: It is HIGHLY recommended that you also
  # enable Gossip encryption.
  # See https://learn.hashicorp.com/consul/security-networking/agent-encryption
  #
  # Note: this relies on functionality introduced with Consul 1.4.1. Make sure
  # your global.image value is at least version 1.4.1.
  tls:
    enabled: true

    # enableAutoEncrypt turns on the auto-encrypt feature on
    # clients and servers.
    # It also switches consul-k8s components to retrieve the CA
    # from the servers via the API.
    # Requires Consul 1.7.1+ and consul-k8s 0.13.0
    enableAutoEncrypt: false

    # serverAdditionalDNSSANs is a list of additional DNS names to
    # set as Subject Alternative Names (SANs) in the server certificate.
    # This is useful when you need to access the Consul server(s) externally,
    # for example, if you're using the UI.
    serverAdditionalDNSSANs: []

    # serverAdditionalIPSANs is a list of additional IP addresses to
    # set as Subject Alternative Names (SANs) in the server certificate.
    # This is useful when you need to access Consul server(s) externally,
    # for example, if you're using the UI.
    serverAdditionalIPSANs: []

    # If verify is true, 'verify_outgoing', 'verify_server_hostname', and
    # 'verify_incoming_rpc' will be set to true for Consul servers and clients.
    # Set this to false to incrementally roll out TLS on an existing Consul cluster.
    # Note: remember to switch it back to true once the rollout is complete.
    # Please see this guide for more details:
    # https://learn.hashicorp.com/consul/security-networking/certificates
    verify: true

    # If httpsOnly is true, Consul will disable the HTTP port on both
    # clients and servers and only accept HTTPS connections.
    httpsOnly: true

    # caCert is a Kubernetes secret containing the certificate
    # of the CA to use for TLS communication within the Consul cluster.
    # If you have generated the CA yourself with the consul CLI,
    # you could use the following command to create the secret in Kubernetes:
    #
    #   kubectl create secret generic consul-ca-cert \
    #           --from-file='tls.crt=./consul-agent-ca.pem'
    #caCert:
    #  secretName: null
    #  secretKey: null

    caCert:
      secretName: consul-ca-cert
      secretKey: crt.key
    # caKey is a Kubernetes secret containing the private key
    # of the CA to use for TLS communications within the Consul cluster.
    # If you have generated the CA yourself with the consul CLI,
    # you could use the following command to create the secret in Kubernetes:
    #
    #   kubectl create secret generic consul-ca-key \
    #           --from-file='tls.key=./consul-agent-ca-key.pem'
    #
    # Note that we need the CA key so that we can generate server and client certificates.
    # It is particularly important for the client certificates since they need to have host IPs
    # as Subject Alternative Names. In the future, we may support bringing your own server
    # certificates.
    caKey:
      secretName: consul-ca-key
      secretKey: tls.key

      #caKey:
      #  secretName: null
      #  secretKey: null
```

gossip-encryption-key is generated manually.

**terraform version** = `12.23`
**helm version** = version.BuildInfo{Version:"v3.2.0", GitCommit:"e11b7ce3b12db2941e90399e874513fbd24bcb71", GitTreeState:"clean", GoVersion:"go1.13.10"}

Expected behavior: helm deployment with tls encryption with certificates generated with terraform.

behavior exibited: init containers crash with error 

`==> WARNING: Server Certificates grants authority to become a
    server and access all state in the cluster including root keys
    and all ACL tokens. Do not distribute them to production hosts
    that are not server nodes. Store them as securely as CA keys.
==> Using /consul/tls/ca/cert/tls.crt and /consul/tls/ca/key/tls.key
unknown PEM block type for signing key: RSA PRIVATE KEY
`
https://github.com/hashicorp/consul/blob/604de8758b94c3686e5907d4402c5b7dc9456c9c/agent/connect/parsing.go#L139

i found this here but the key is valid and it has an option for rsa private key. hope this is enough info for you guys! thanks in advance!

**Comment by lkysow** [comment]
Hi @Epiclemonaid it looks like RSA keys aren't supported: https://github.com/hashicorp/consul/issues/7622. The code you found is for Consul Connect which is a different subsytem from the server certificates. You'll need to follow that issue for updates and until it's supported use EC keys.

**Comment by ishustava** [comment]
Hey @Epiclemonaid,

To add to what Luke said unless you have a specific requirement to only use RSA keys, I'd recommend just changing your terraform code to use ECDSA keys:

```
resource "tls_private_key" "example" {
  algorithm   = "ECDSA"
  ecdsa_curve = "P384"
}
```

There's a number of benefits to using ECDSA keys, for example, that the key sizes and the associated certificates are smaller.

Also, a side note, you don't need to generate Consul server certificates through terraform if you are setting `global.tls.caCert` and `global.tls.caKey`. The Helm chart will generate server certs for you and store them as secrets. 

If your intention is to generate server and client certificates yourself, unfortunately, we don't have full support for this at the moment. If you are interested in this being supported, we'd love to know!

**Comment by ishustava** [comment]
Hi @Epiclemonaid 

I'm going to close this issue because it's out of scope for this repo. I'd recommend either using the ECDSA keys or following the Consul issue Luke has mentioned for updates if you require RSA keys. If you feel that this issue is unresolved, please let us know!

**Comment by getjack** [comment]
Hi @ishustava 
i just ran into the same problem. I want to reuse an existing cert + private key which is RSA 2048 but as i can see under /consul
/tlsutil/generate.go only ECDSA is supported.

A little overview over my setup:
We own an existing consul (server) cluster and i now want to join a few clients on K8S via the helm deployment. As my requirement is to use the clients only for internal communication (which is secured network-wise) but demand https i want to reuse my wildcard certificate which is in use on the existing cluster.

As the certificate is already available in my K8S cluster i thought i could just add the secret to the vaules.yaml but the pods won't initialize. Also i was only able to debug this because we have an ELK stack collection K8S logs as the pods would not print any error message fast enough for me to look into.

I am not an expert regarding certificates an my work colleagues are on vaccation but i doubt i can convert those certs.

**Comment by ishustava** [comment]
Hey @getjack,

Thanks for sharing your use case. The best way to fix this is to share this in the hashicorp/consul#7622 since it's the issue with Consul itself and isn't in-scope for consul-helm to solve.

One way you can work around this in your situation is to use auto-encrypt. This feature allows clients to ask servers to issue a certificate on start-up. Under the hood, Consul client certificates will be provisioned with the same CA as the one used for Consul Connect. Note that you need [`connect`](https://www.consul.io/docs/agent/options#connect_enabled) to be enabled on your Consul servers for this to work. 

You could read more about it in our docs on [how to configure external Consul servers](https://www.consul.io/docs/k8s/installation/deployment-configurations/servers-outside-kubernetes#configuring-tls-with-auto-encrypt).

**Comment by ashwinkupatkar** [comment]
Hi @ishustava,

I am running into a similar problem. (slightly different) 

I created a set of ECDSA private key and its CSR . I signed the CSR with the intermediate from vault. I got the signed CA certificate from vault following this document: https://www.vaultproject.io/api-docs/secret/pki#sign-intermediate

I created a chain as follows:

[ generated intermediate CA cert3(ecdsa) -> intermediate CA cert2(rsa) -> intermediate CA cert1(rsa) -> root cert(rsa) ]
 
Now I am using this **generated intermediate CA cert3(ecdsa)** to create a Consul CA with below command

FYI generated intermediate CA cert3(ecdsa),

consul CA cert name: consul_root_ca.pem
consul CA key name: consul_root_ca_key.pem

`consul tls cert create -server -node=<nodename> -dc=dc1 -ca=consul_root_ca_key.pem -key=consul_root_ca_key.pem`

Output:

`==> WARNING: Server Certificates grants authority to become a
    server and access all state in the cluster including root keys
    and all ACL tokens. Do not distribute them to production hosts
    that are not server nodes. Store them as securely as CA keys.
==> Using consul_root_ca_key.pem and consul_root_ca_key.pem
**first PEM-block should be CERTIFICATE type`**

How can this be fixed ?

I thought of generating an RSA private key and its CSR for my **generated intermediate CA cert3** .... but looks from the above discussion they aren't supported.

what can be done in this scenario ?

**Comment by ashwinkupatkar** [comment]
Hi @ishustava,

When we use auto-encrypt, how will the envoy proxy consume the dynamic certs provided to consul client.

Currently if the TLS is enabled in the consul cluster. we need to pass below environment variables
CONSUL_CACERT, CONSUL_CLIENT_CERT and CONSUL_CLIENT_KEY

In this scenario where i donot know the consul client certs and if I just provide the CONSUL_CACERT it complains of **tls:bad certificate** or **certificate signed by unknown authority**

what can be done here ?

**Comment by ishustava** [comment]
Hey @ashwinkupatkar 

It looks like in the situation you describing, you're using mTLS, i.e. have `verify_incoming` or `verify_incoming_https` set to true. If you're using auto-encrypt, you should turn off mTLS for **HTTPS** connections (set `verify_incoming_https` to false) for the **clients** and instead, rely on ACLs for authentication. In that case, you don't need to provide client cert/key. Please take a look at [this guide](https://learn.hashicorp.com/tutorials/consul/tls-encryption-secure#configure-the-clients). With regard to CA for the clients, you can get it from [this endpoint](https://www.consul.io/api-docs/connect/ca#list-ca-root-certificates). 

If you have other questions, please open a new issue as this is quite different from this original issue.

**Comment by ashwinkupatkar** [comment]
Thankyou @ishustava . Let me take a look on it. 

Sure for issue persisting around the same issue that i mentioned, will open up a separate ticket.
